modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/transaction/ContextualTransactionValidator.scala

```diff
@@ -0,0 +1,265 @@
+        lastProcessedTransactionRef: TransactionReference
+      ): Either[ContextualTransactionValidationError, Hashed[Transaction]] =
+        if (transaction.parent.ordinal < lastProcessedTransactionRef.ordinal)
+          ParentOrdinalLowerThenLastProcessedTxOrdinal(transaction.parent.ordinal, lastProcessedTransactionRef.ordinal).asLeft
+        else if (transaction.parent === lastProcessedTransactionRef) {
+          transaction.asRight
+        } else {
+          val hasValidParent = txs.exists(_.exists {
+            case (parentOrdinal, parentTx) => parentTx.ref === transaction.parent && parentOrdinal > lastProcessedTransactionRef.ordinal
+          })
+
+          if (hasValidParent) transaction.asRight else HasNoMatchingParent(transaction.parent.hash).asLeft
+        }
+
+      private def validateNotLocked(
+        transaction: Hashed[Transaction]
+      ): Either[ContextualTransactionValidationError, Hashed[Transaction]] =
+        Either
+          .cond(
+            !lockedAddresses.contains(transaction.source),
+            transaction,
+            LockedAddressError(transaction.source)
+          )
+
+      private def calculateLastAllowedTransactionWithinCap(
+        transactions: List[Hashed[Transaction]],
+        balance: Balance,
+        lastMajorityTransactionOrdinal: SnapshotOrdinal,
+        currentOrdinal: SnapshotOrdinal
+      ): Option[Hashed[Transaction]] = {
+        val ordinalDistanceForBaseBalance = Math
+          .floor(transactionLimitConfig.timeToWaitForBaseBalance / transactionLimitConfig.timeTriggerInterval)
+          .toLong
+
+        @tailrec
+        def loop(
+          remainingTransactions: List[Hashed[Transaction]],
+          balance: Balance,
+          lastTxRefOrdinal: SnapshotOrdinal,
+          totalConsumedCap: Double,
+          lastAcceptedWithinCap: Option[Hashed[Transaction]]
+        ): Option[Hashed[Transaction]] =
+          remainingTransactions match {
+            case head :: tail =>
+              val consumedCap =
+                if (head.fee.value >= transactionLimitConfig.minFeeWithoutLimit.value) 0
+                else {
+                  val distancePassed = currentOrdinal.value - lastTxRefOrdinal.value
+                  val transactionsAllowedWithinInterval =
+                    if (balance.value > 0) balance.value.toDouble / transactionLimitConfig.baseBalance.value.toDouble else 0.0
+                  val distanceBetweenTxsNeeded = ordinalDistanceForBaseBalance.toDouble / transactionsAllowedWithinInterval
+                  distanceBetweenTxsNeeded / Math.max(distancePassed, 1)
+                }
+
+              val updatedTotalConsumedCap = totalConsumedCap + consumedCap
+              val maybeRemainingBalance = toAmount(head.amount).plus(head.fee).flatMap(balance.minus)
+
+              maybeRemainingBalance match {
+                case Right(remainingBalance) if updatedTotalConsumedCap <= 1 && remainingBalance.value > 0 && tail.nonEmpty =>
```

Comment:
I think you could drop the last two conditions and just let it execute once more, then the second case will not be required and on a subsequent execution the default case will be triggered for balance = 0 or if there were no more transactions.